### PR TITLE
Docker: add LTP test scripts

### PR DESCRIPTION
**Pull request by llugin**
Changes:
- refactor of existing Sqlite scripts to make code reusable by LTP
- elapsed time of single test case and whole execution now shown
- changes in data structure for storing ran test results
- handling for possible errors in utf-8 decoding of test outputs
- tests that stopped failing compared to fails text file now listed (instead of 'some tests stopped failing')
- LTP test scripts added
- LTP integrated with docker

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/324)

### CODE REVIEW

**Comment by marcinslusarz** [comment]
Reviewed 13 of 28 files at r3, 1 of 1 files at r4.
Review status: 11 of 26 files reviewed at latest revision, 4 unresolved discussions.

---

*[utils/docker/external_tests/suite.py, line 59 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtH6xHylM17p0EaK4EU:-KtH6xHylM17p0EaK4EV:bax5t4s) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/external_tests/suite.py#L59)):*
> ```Python
>         results = ['{0:>15} on {1} {2:>{3}.3f} [ms]'.
>                    format(self.test_entry[config]['result'], config,
>                           self.test_entry[config]['time'], 35 - len(config))
> ```

Where is 35 coming from?

---

*[utils/docker/external_tests/ltp/__init__.py, line 0 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtH60O3QsvzlG4QCJhF:-KtH60O4JVtnNNFesdJl:b-q6cqpx) ([raw file](https://github.com/pmem/pmemfile/blob/01eeaa0f04116e5ac6a4b90ebff6a3ddc1d3dfc3/utils/docker/external_tests/ltp/__init__.py#L0)):*
Do we need empty file in repo?

---

*[utils/docker/images/Dockerfile.fedora-23, line 94 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtGz2jzgvJAPLlF6eRi:-KtGz2jzgvJAPLlF6eRj:b-awzib3) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/images/Dockerfile.fedora-23#L94)):*
> ```
> RUN gpasswd wheel -a $USER
> 
> USER $USER
> ```

Why do you switch users during installation?

---

*[utils/docker/images/install-ltp.sh, line 2 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtGyoudT5dogDbQ19vg:-KtGyoudT5dogDbQ19vh:b-4w3l4i) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/images/install-ltp.sh#L2)):*
> ```Shell
> #!/bin/bash -xe
> ```

Missing copyright and license.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by GBuella** [comment]
Review status: 11 of 26 files reviewed at latest revision, 4 unresolved discussions.

---

*[utils/docker/external_tests/ltp/__init__.py, line  at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtH60O3QsvzlG4QCJhF:-KtJ49t4a5xx_Vn7_oT4:b-k1lck9) ([raw file](https://github.com/pmem/pmemfile/blob/01eeaa0f04116e5ac6a4b90ebff6a3ddc1d3dfc3/utils/docker/external_tests/ltp/__init__.py#L)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Do we need empty file in repo?
</blockquote></details>

The copyright statement is missing from this file!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by sarahjelinek** [comment]
Review status: 11 of 26 files reviewed at latest revision, 7 unresolved discussions.

---

*[utils/docker/external_tests/run-build-suite.sh, line 75 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO3eMuC5ReV5e7rRg3:-KtO3eMuC5ReV5e7rRg4:bpw1b6j) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/run-build-suite.sh#L75)):*
> ```Shell
> 	SUITE_UTILS_DIR=$TEST_UTILS_DIR/ltp
> else
> 	echo "First argument doesn't match any of existing suites"\
> ```

s/any of existing suites/any existing test suites

---

*[utils/docker/external_tests/tester.py, line 94 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO4ibKRl7nmeAbK6n8:-KtO4ibKRl7nmeAbK6n9:bukyzlx) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/tester.py#L94)):*
> ```Python
> 
>         if set(self.failed_pf_only) == set(past_fails):
>             print('No new fails.')
> ```

s/No new fails/No new failed tests

---

*[utils/docker/external_tests/tester.py, line 100 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO4nW0au8lMyjyVIyk:-KtO4nW0au8lMyjyVIyl:b-46qo1w) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/tester.py#L100)):*
> ```Python
>         if new_fails:
>             exit_code = 1
>             print('New fails introduced in this execution:')
> ```

s/New fails/New test failures

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by krzycz** [comment]
Reviewed 11 of 22 files at r1, 28 of 28 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by llugin** [comment]
Review status: 13 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[utils/docker/external_tests/run-build-suite.sh, line 75 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO3eMuC5ReV5e7rRg3:-KtkuvLaVDOEcPCozrUj:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/run-build-suite.sh#L75)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

s/any of existing suites/any existing test suites
</blockquote></details>

Done.

---

*[utils/docker/external_tests/suite.py, line 59 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtH6xHylM17p0EaK4EU:-Ktkv-JXQwqFBKBLbanU:bwldnv1) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/external_tests/suite.py#L59)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Where is 35 coming from?
</blockquote></details>

assigned 35 to 'config_margin' variable

---

*[utils/docker/external_tests/tester.py, line 94 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO4ibKRl7nmeAbK6n8:-KtkvBJlfUvZJYeAiP5r:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/tester.py#L94)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

s/No new fails/No new failed tests
</blockquote></details>

Done.

---

*[utils/docker/external_tests/tester.py, line 100 at r4](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtO4nW0au8lMyjyVIyk:-KtkvBwmemHdjA4h1Ydi:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/ed40ab34af8bbc46f297b2e7b1e037aada2489b7/utils/docker/external_tests/tester.py#L100)):*
<details><summary><i>Previously, sarahjelinek (Sarah Jelinek) wrote…</i></summary><blockquote>

s/New fails/New test failures
</blockquote></details>

Done.

---

*[utils/docker/external_tests/ltp/__init__.py, line  at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtH60O3QsvzlG4QCJhF:-KtkvD3nMwr_iJc049Ki:b-2aut3y) ([raw file](https://github.com/pmem/pmemfile/blob/01eeaa0f04116e5ac6a4b90ebff6a3ddc1d3dfc3/utils/docker/external_tests/ltp/__init__.py#L)):*
>Do we need empty file in repo?
Yes.

---

*[utils/docker/images/Dockerfile.fedora-23, line 94 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtGz2jzgvJAPLlF6eRi:-KtkvIxDBfl_mgktfvAQ:bc161la) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/images/Dockerfile.fedora-23#L94)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Why do you switch users during installation?
</blockquote></details>

Fixed, ltp now installed by root.

---

*[utils/docker/images/install-ltp.sh, line 2 at r3](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KtGyoudT5dogDbQ19vg:-KtkvOi3r5aFq3B4nQCe:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/df97417261a514f8723ede4240020bfcb6526a57/utils/docker/images/install-ltp.sh#L2)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Missing copyright and license.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by krzycz** [comment]
Reviewed 1 of 28 files at r3, 13 of 13 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[utils/docker/external_tests/run-build-suite.sh, line 107 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ktl0XuQiVnXpQLVNtGC:-Ktl0XuQiVnXpQLVNtGD:b-94uxvc) ([raw file](https://github.com/pmem/pmemfile/blob/1f669a08258a106d290a5dc6c1b187fa680a621f/utils/docker/external_tests/run-build-suite.sh#L107)):*
> ```Shell
> 
> 
> EXIT_CODE=`echo $?`
> ```

Is echo required here?
`EXIT_CODE=$?`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by llugin** [comment]
Review status: 15 of 26 files reviewed at latest revision, 8 unresolved discussions.

---

*[utils/docker/external_tests/run-build-suite.sh, line 107 at r5](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ktl0XuQiVnXpQLVNtGC:-KtlJcdm8wHyO_cYkldj:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/1f669a08258a106d290a5dc6c1b187fa680a621f/utils/docker/external_tests/run-build-suite.sh#L107)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is echo required here?
`EXIT_CODE=$?`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by krzycz** [comment]
Reviewed 2 of 28 files at r3, 11 of 11 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

File: utils/docker/external_tests/run-build-suite.sh [hunk 1]
```diff
@@ -68,24 +92,28 @@ rm -r *
...
-	--timeout 120
+$TEST_UTILS_DIR/run-suite.py $1 -i $SUITE_DIR -m $MOUNTPOINT -l $PMEMFILE_LIB_DIR \
+	-p $PF_POOL -t $TESTS -f $FAILING_TESTS --timeout 120 $VERBOSE
+
+
+EXIT_CODE=$?
 
 if [ "$COVERAGE" = "1" ]; then
-	bash <(curl -s https://codecov.io/bash) -c -F sqlite_tests
+	bash <(curl -s https://codecov.io/bash)
```
Comment (marcinslusarz) [inline utils/docker/external_tests/run-build-suite.sh]: Don't force your version when resolving conflicts!

**Comment by marcinslusarz** [comment]
Reviewed 2 of 28 files at r3, 7 of 11 files at r6.
Review status: 25 of 26 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by llugin** [comment]
Review status: 14 of 26 files reviewed at latest revision, 5 unresolved discussions.

---

*[utils/docker/external_tests/run-build-suite.sh, line 110 at r6](https://reviewable.io:443/reviews/pmem/pmemfile/324#-KsxUNx09ROLtUm1r6aU-r6-110:-KtuyvOpOt_2vUFC7Cz2:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/fac667c7b87720592efef922afe6cbba98232af7/utils/docker/external_tests/run-build-suite.sh#L110)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Don't force your version when resolving conflicts!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by GBuella** [comment]
Reviewed 6 of 28 files at r3, 1 of 13 files at r5, 1 of 1 files at r7, 8 of 9 files at r8, 3 of 3 files at r9.
Review status: 25 of 26 files reviewed at latest revision, 5 unresolved discussions.

---

*[utils/docker/external_tests/config.py, line 36 at r9](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ktv8GzTDh1nUj8PgIW5:-Ktv8GzUxU6k9km23Lfr:b9rfztn) ([raw file](https://github.com/pmem/pmemfile/blob/892433a2a34dcb18249b92d5a99f97582d3ad027/utils/docker/external_tests/config.py#L36)):*
> ```Python
> 
> 
> class Config:
> ```

I'm far from being a Python expert... what is the point of this class?
Wouldn't basically an associative array be able to do the same job?
Or, a function returning an associative array?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by llugin** [comment]
Review status: 25 of 26 files reviewed at latest revision, 5 unresolved discussions.

---

*[utils/docker/external_tests/config.py, line 36 at r9](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ktv8GzTDh1nUj8PgIW5:-KtvB1uz2AyBi89FTLWj:bss7cb3) ([raw file](https://github.com/pmem/pmemfile/blob/892433a2a34dcb18249b92d5a99f97582d3ad027/utils/docker/external_tests/config.py#L36)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

I'm far from being a Python expert... what is the point of this class?
Wouldn't basically an associative array be able to do the same job?
Or, a function returning an associative array?
</blockquote></details>

I found class properties to be useful here.

1. I need pf_env to be calculated dynamically on access.  It has to contain all previously set env variables and since these env vars may be set by other parts of the python script (such thing happen with ltp script) it has to be up-to-date.

2.  To prevent mountpoint dir to be changed outside initialization it has only getter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by sarahjelinek** [comment]
Reviewed 1 of 28 files at r3, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by krzycz** [comment]
Reviewed 2 of 28 files at r3, 1 of 1 files at r7, 8 of 9 files at r8, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by GBuella** [comment]
Reviewed 5 of 28 files at r3, 1 of 11 files at r6, 1 of 3 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Reviewed 1 of 22 files at r2, 5 of 28 files at r3, 4 of 13 files at r5, 1 of 11 files at r6, 1 of 1 files at r7, 8 of 9 files at r8, 1 of 3 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[utils/docker/external_tests/run-build-suite.sh, line 46 at r11](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ku-IaB2yLjFuQ4Xkgvy:-Ku-IaB3XCaTX8VUb9H5:b3kry) ([raw file](https://github.com/pmem/pmemfile/blob/de2d5fa6680a278a0b05747f1fa7e37379b2682c/utils/docker/external_tests/run-build-suite.sh#L46)):*
> ```Shell
> if [[ -z "$1" ]]; then
> 	echo "ERROR: First argument has to contain a name of test suite"\
> 	"to be run (sqlite|ltp)."
> ```

S

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324#-:-Ku01vhbRTMkB0Sw9DM1:bnfp4nl)*

**Comment by marcinslusarz** [comment]
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[utils/docker/external_tests/run-build-suite.sh, line 46 at r11](https://reviewable.io:443/reviews/pmem/pmemfile/324#-Ku-IaB2yLjFuQ4Xkgvy:-Ku02Lt1fS_cMaq4hMsa:bltau3o) ([raw file](https://github.com/pmem/pmemfile/blob/de2d5fa6680a278a0b05747f1fa7e37379b2682c/utils/docker/external_tests/run-build-suite.sh#L46)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

S
</blockquote></details>

doesn't matter

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by krzycz** [comment]
Reviewed 1 of 28 files at r3, 2 of 2 files at r11.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by marcinslusarz** [comment]
Reviewed 2 of 2 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Comment by GBuella** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324#-:-Ku4c5a35RFspaP4WlOE:bnfp4nl)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324#-:-KuJ_UEUPRwWcIqmq-tG:bnfp4nl)*

**Comment by krzycz** [comment]
Reviewed 1 of 2 files at r12.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/324)*

**Pull request merged**